core/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -0,0 +1,361 @@
+                String currentFieldName = parser.currentName();
+                if (SNAPSHOTS.equals(currentFieldName)) {
+                    if (parser.nextToken() == XContentParser.Token.START_ARRAY) {
+                        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                            snapshots.add(SnapshotId.fromXContent(parser));
+                        }
+                    } else {
+                        throw new ElasticsearchParseException("expected array for [" + currentFieldName + "]");
+                    }
+                } else if (INDICES.equals(currentFieldName)) {
+                    if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
+                        throw new ElasticsearchParseException("start object expected [indices]");
+                    }
+                    while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        String indexName = parser.currentName();
+                        String indexId = null;
+                        Set<SnapshotId> snapshotIds = new LinkedHashSet<>();
+                        if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
+                            throw new ElasticsearchParseException("start object expected index[" + indexName + "]");
+                        }
+                        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                            String indexMetaFieldName = parser.currentName();
+                            parser.nextToken();
+                            if (INDEX_ID.equals(indexMetaFieldName)) {
+                                indexId = parser.text();
+                            } else if (SNAPSHOTS.equals(indexMetaFieldName)) {
+                                if (parser.currentToken() != XContentParser.Token.START_ARRAY) {
+                                    throw new ElasticsearchParseException("start array expected [snapshots]");
+                                }
+                                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                                    snapshotIds.add(SnapshotId.fromXContent(parser));
+                                }
+                            }
+                        }
+                        assert indexId != null;
+                        indices.put(indexName, new IndexMeta(indexName, indexId, snapshotIds));
+                    }
+                } else {
+                    throw new ElasticsearchParseException("unknown field name  [" + currentFieldName + "]");
+                }
+            }
+        } else {
+            throw new ElasticsearchParseException("start object expected");
+        }
+        return new RepositoryData(snapshots, indices);
+    }
+
+    /**
+     * Represents information about a single index snapshotted in a repository.
+     */
+    public static final class IndexMeta implements Writeable {
```

Comment:
You're right, it doesn't need to be Writeable. I'm up for brainstorming a better representation and/or name for this structure.